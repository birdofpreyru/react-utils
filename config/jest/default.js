/* eslint-disable import/no-commonjs */
/* global __dirname, module, process, require */

const path = require('node:path');

const globalLibDir = path.resolve(process.execPath, '../../lib/node_modules');

module.exports = {
  collectCoverage: true,
  collectCoverageFrom: [
    'src/**/*.{js,jsx,ts,tsx}',
  ],
  coverageDirectory: '__coverage__',
  moduleDirectories: [
    'node_modules',
    globalLibDir,
  ],
  moduleNameMapper: {
    '\\.(scss|css)$': 'identity-obj-proxy',
  },
  modulePathIgnorePatterns: ['<rootDir>/build/'],
  resolver: `${__dirname}/resolver.js`,
  rootDir: '../..',
  setupFilesAfterEnv: [
    '<rootDir>/config/jest/setup.js',
  ],
  testEnvironmentOptions: {
    url: 'http://localhost',
  },
  testMatch: ['**/__tests__/**/*.(j|t)s?(x)'],
  testPathIgnorePatterns: [
    '/__assets__/',
    '/node_modules/',
  ],
  transform: {
    '\\.((j|t)sx?|svg)$': 'babel-jest',
  },
  transformIgnorePatterns: [
    '/node_modules/(?!@dr.pogodin/react-utils|uuid)',
  ],

};
