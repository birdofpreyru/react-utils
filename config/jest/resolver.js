/* global module */

module.exports = (path, options) => {
  const ops = { ...options };

  // Appending "development" option we ensure that Jest tests use development
  // version of the library's build for browsers, which for example contains
  // "data-testid" attributes (they are optimized-out of production builds).
  if (ops.conditions && !ops.conditions.includes('development')) {
    ops.conditions = [...ops.conditions, 'development'];
  }

  return options.defaultResolver(path, ops);
};
