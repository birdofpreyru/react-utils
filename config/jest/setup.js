/* global globalThis, require */

// eslint-disable-next-line import/no-unassigned-import
import 'raf/polyfill';

// TODO: This is a temporary polyfill necessary for react-router,
// as JSDom does not provide TextEncoder, see:
// https://github.com/remix-run/react-router/issues/12363
if (!globalThis.TextEncoder || !globalThis.TextDecoder) {
  const { TextDecoder, TextEncoder } = require('node:util');
  globalThis.TextEncoder = TextEncoder;
  globalThis.TextDecoder = TextDecoder;
}
