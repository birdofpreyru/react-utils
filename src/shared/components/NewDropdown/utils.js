/**
 * Auxiliary functions for <Dropdown> component.
 */

import { findIndex, findLastIndex } from 'lodash';

/**
 * Returns option name and value as two elements of array.
 * @param {object|string} option
 * @return {string[]}
 */
export function optionNameValue(option) {
  return typeof option === 'string'
    ? [option, option] : [option.name, option.value];
}

/**
 * Returns option value.
 * @param {object|string} option Option object or string.
 * @return {string}
 */
export function optionValue(option) {
  return typeof option === 'string' ? option : option.value;
}

/**
 * Finds the option by value.
 * @param {string} value Option value.
 * @param {array} options Options array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Index of the matching option, or -1.
 */
function findOptionIndex(value, options, filter) {
  return options.findIndex(
    (item) => optionValue(item) === value && (!filter || filter(item)),
  );
}

/**
 * Finds the next option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Next option index, or -1.
 */
export function findNextOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter);
  if (index >= 0 && ++index < options.length) {
    return filter ? findIndex(options, filter, index) : index;
  }
  return -1;
}

/**
 * Finds the previous option index.
 * @param {string} value Current value.
 * @param {array} options Option array.
 * @param {function} [filter] Optional. Filter function.
 * @return {number} Previous option index, or -1.
 */
export function findPrevOptionIndex(value, options, filter) {
  let index = findOptionIndex(value, options, filter) - 1;
  if (index >= 0 && filter) index = findLastIndex(options, filter, index);
  return index >= 0 ? index : -1;
}
