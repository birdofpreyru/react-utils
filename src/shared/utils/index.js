import themed, {
  COMPOSE,
  PRIORITY,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as url from './url';
import * as webpack from './webpack';

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

let JU = null; // eslint-disable-line import/no-mutable-exports
if (isomorphy.IS_SERVER_SIDE
  && (process.env.NODE_CONFIG_ENV || process.env.NODE_ENV) !== 'production'
) {
  /* eslint-disable global-require */
  const path = webpack.requireWeak('path');
  JU = webpack.requireWeak(path.resolve(__dirname, './jest'));
  /* eslint-enable global-require */
}

/**
 * Creates a new async barrier: a Promise instance with its resolve method
 * attached as .resolve field.
 * @return {Promise<>}
 */
function newBarrier() {
  let resolve;
  const promise = new Promise((res) => { resolve = res; });
  promise.resolve = resolve;
  return promise;
}

/**
 * Attempts to execute provided async "action" up to "maxRetries" time with
 * the given "interval". If any of attempts is successful it returns its result,
 * otherwise if all attempts fail it throws the error of last attempt.
 * @param {function} action
 * @param {number} [maxRetries=5] Optional. Maximum number of retries. Defaults
 *  to 5 attempts.
 * @param {number} [interval=1000] Optional. Interval between retries [ms].
 *  Defaults to 1 second.
 * @return {Promise} Resolves to the result of successful operation, or
 *  rejects with the error from the latst failed attempt.
 */
async function withRetries(action, maxRetries = 5, interval = 1000) {
  /* eslint-disable no-await-in-loop */
  for (let n = 1; ; ++n) {
    try {
      return await action();
    } catch (error) {
      if (n < maxRetries) await time.timer(interval);
      else throw error;
    }
  }
  /* eslint-enable no-await-in-loop */
}

export {
  config,
  isomorphy,
  JU,
  newBarrier,
  themed,
  ThemeProvider,
  time,
  url,
  webpack,
  withRetries,
};
