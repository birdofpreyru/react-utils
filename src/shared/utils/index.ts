import themedImpl, {
  COMPOSE,
  PRIORITY,
  type Theme,
  ThemeProvider,
} from '@dr.pogodin/react-themes';

import config from './config';
import * as isomorphy from './isomorphy';
import time from './time';
import * as webpack from './webpack';

export {
  type Listener,
  Barrier,
  Emitter,
  Semaphore,
  withRetries,
} from '@dr.pogodin/js-utils';

export { getSsrContext } from './globalState';
export { default as splitComponent } from './splitComponent';

type ThemedT = typeof themedImpl & {
  COMPOSE: typeof COMPOSE;
  PRIORITY: typeof PRIORITY;
};

const themed: ThemedT = themedImpl as ThemedT;

themed.COMPOSE = COMPOSE;
themed.PRIORITY = PRIORITY;

export {
  type Theme,
  config,
  isomorphy,
  themed,
  ThemeProvider,
  time,
  webpack,
};
