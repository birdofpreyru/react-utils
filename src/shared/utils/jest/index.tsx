/* global jest, document */
/* eslint-disable import/no-extraneous-dependencies */

import type {
  AxiosRequestConfig,
  AxiosResponse,
  AxiosStatic,
  InternalAxiosRequestConfig,
} from 'axios';

import mockdate from 'mockdate';
import { type ReactNode, act } from 'react';
import { type Root, createRoot } from 'react-dom/client';

import { type RenderResult, render } from '@testing-library/react';

/**
 * An alias for [act(..)](https://reactjs.org/docs/test-utils.html#act)
 * from `react`.
 */
export { act };

export { default as getGlobal } from './global';

global.IS_REACT_ACT_ENVIRONMENT = true;

const originalProcessVersions = process.versions;

/**
 * Tricks **react-utils** into thinking the test is running within client-side
 * (browser) environment.
 */
export function mockClientSide() {
  Object.defineProperty(process, 'versions', { value: undefined });
}

/**
 * Reverts the effect of {@link module:JU.mockClientSide mockClientSide(..)}.
 */
export function unmockClientSide() {
  Object.defineProperty(process, 'versions', {
    value: originalProcessVersions,
    writable: false,
  });
}

/**
 * Generates a mock UUID, or better said it determenistically transforms given
 * `seed` number into a UUID-formatted string.
 * @param {number} seed
 * @return {string}
 */
export function getMockUuid(seed = 0) {
  const x = seed.toString(16).padStart(32, '0');
  return `${x.slice(0, 8)}-${x.slice(8, 12)}-${x.slice(12, 16)}-${x.slice(16, 20)}-${x.slice(20)}`;
}

export type AxiosRequestHandlerT =
  (config: AxiosRequestConfig) => Partial<AxiosResponse> | null | undefined;

export function mockAxios(handlers: AxiosRequestHandlerT[]) {
  const axios: AxiosStatic = jest.requireActual('axios');

  axios.defaults.adapter = async (
    config: AxiosRequestConfig,
  ): Promise<AxiosResponse> => {
    for (const handler of handlers) {
      const res = handler(config);
      if (res) {
        return {
          config: config as InternalAxiosRequestConfig,
          data: null,
          headers: {},
          status: 200,
          statusText: 'OK',
          ...res,
        };
      }
    }

    // Fallback to the regular network request.
    let res: AxiosResponse;
    try {
      res = await axios({ ...config, adapter: ['xhr', 'http', 'fetch'] });
      console.warn(
        'Network request has not been mocked for a test.\n\nConfig:\n',
        config,
        '\n\nResult:\n',
        JSON.stringify(res, null, 2),
      );
    } catch (e) {
      console.warn(
        'Network request has not been mocked for a test, and failed.\n\nConfig:\n',
        config,
        '\n\nError\n',
        JSON.stringify(e, null, 2),
      );
      throw e;
    }

    return res;
  };

  return axios;
}

/**
 * Advances mock timers, and mock date by the specified time.
 * @param {number} time Time step [ms].
 * @returns {Promise} Wait for this to "jump after" any async code which should
 * be executed because of the mock time movement.
 */
export async function mockTimer(time: number) {
  mockdate.set(time + Date.now());
  await jest.advanceTimersByTimeAsync(time);
}

export type MountedSceneT = HTMLElement & {
  destroy: () => void;
  snapshot: () => void;
};

/**
 * Mounts `scene` to the DOM, and returns the root scene element.
 * @param scene
 * @return Created container DOM element with destroy() function
 *  attached.
 */
export function mount(scene: ReactNode): MountedSceneT {
  let root: Root;
  const element = document.createElement('div');
  document.body.appendChild(element);

  const res: MountedSceneT = (element as unknown) as MountedSceneT;
  res.destroy = () => {
    // NOTE: As it seems @testing-library may reset this flag to false
    // when it is simulating user events.
    global.IS_REACT_ACT_ENVIRONMENT = true;

    act(() => root.unmount());
    res.remove();
  };

  res.snapshot = () => {
    expect(res).toMatchSnapshot();
  };

  // NOTE: As it seems @testing-library may reset this flag to false
  // when it is simulating user events.
  global.IS_REACT_ACT_ENVIRONMENT = true;

  act(() => {
    root = createRoot(res);
    root.render(scene);
  });
  return res;
}

type SnapshotOptionsT = {
  await?: Promise<void>;
};

export async function snapshot(
  element: React.ReactElement,
  options?: SnapshotOptionsT,
) {
  let res: RenderResult | undefined;

  const promise = act(() => {
    res = render(element);
    return options?.await;
  });

  if (res === undefined) throw Error('Render failed');
  if (options?.await) await promise;

  const nodes = res.asFragment().childNodes;
  expect(nodes.length > 1 ? [...nodes] : nodes[0]).toMatchSnapshot();
  return res;
}
