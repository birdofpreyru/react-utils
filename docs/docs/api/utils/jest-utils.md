# Jest Utils
Collection of helpers useful in [Jest] tests.

:::caution
Prior to **v1.35.0** these utilities were exported as `JU` module of the main
package export, now they are exposed via a dedicated secondary export.

<details>
For example, to use [snapshot()] helper in a test prior to **v1.35.0** you would
do (and `JU` would be `null` when imported in non-`development` build, or outside
the server-side environment):
```tsx
import { JU } from '@dr.pogodin/react-utils';

test('Example', () => {
  JU!.snapshot(<div>Hello World</div>);
});
```
Starting with **v1.35.0** you must do:
```tsx
/** @jest-environment jsdom */

import { snapshot } from '@dr.pogodin/react-utils/jest';

test('Example', () => {
  snapshot(<div>Hello World</div>);
});
```
The change was done to avoid module resolution issues when using JSDom
environment for Jest tests.
</details>
:::

## E2eSsrEnv
Although it is not exported from this module for technical reasons, [E2eSsrEnv]
class implements Jest environment for end-to-end testing of SSR and client-side
execution of Webpack-built code with Jest, and thus should be considered as
a very important part of Jest-testing utilities provided by the library.

## Methods
- [act()](#act) &mdash; An alias for
  the [act()](https://reactjs.org/docs/test-utils.html#act) function from
  `react` (moved into `react` package since React v18.3).
- [getMockUuid()](#getmockuuid) - Generates a mock UUID (formats the given `seed`
  into an UUID-formatted string).
- [mockAxios()] &mdash; A helper which simplifies mocking out of [axios]
  (our selected HTTP(S) client library) in tests.
- [mockClientSide()] - Tricks **react-utils** into thinking the test is running
  within the client-side (browser) environment.
- [mockTimer()](#mocktimer) - Advances mock timers, and mock date by
  the specified time step.
- [mount()](#mount) - Mounts `scene` to the DOM, and returns the root scene
  element with **.destroy()** method attached.
- [snapshot()](#snapshot) - Does does a snapshot test of the give ReactJS
  component, and also returns JSON representation of is render.
- [unmockClientSide()](#unmockclientside) - Reverts the effect of previous
  [mockClientSide()] call.

:::caution Deprecated Methods
- [render()](#render) &mdash; Removed in **v1.34.0**. Migrate to
  [render()] provided by [@testing-library/react].
- [shallowRender()] &mdash; Removed in **v1.34.0**. Starting with React v19
  it is recommended to [avoid shallow rendering] in tests.
- [shallowSnapshot()](#shallowsnapshot) &mdash; Removed in **v1.34.0**.
  Starting with React v19 it is recommended to [avoid shallow rendering]
  in tests.
- [simulate](#simulate) &mdash; Removed in **v1.34.0**. Use instead
  [@testing-library/user-event], or the lower-level [fireEvent] from
  [@testing-library/dom].
:::

### act()
```tsx
import { act } from '@dr.pogodin/react-utils/jest';
```
This method is just an alias for
[act()](https://reactjs.org/docs/test-utils.html#act) function from
`react` (moved into `react` since v18.3).

### getMockUuid()
```tsx
import { getMockUuid } from '@dr.pogodin/react-utils/jest';

function getMockUuid(seed = 0): string;
```
Generates a mock UUID by determenistically transforming the given `seed` into
a UUID-formatted string.

**Arguments**
- `seed` &mdash; **number** &mdash; Optional. Defaults `0`.

**Returns**
- **string** - Mock UUID.

### mockAxios()
[mockAxios()]: #mockaxios
```ts
import { mockAxios } from '@dr.pogodin/react-utils/jest';

function mockAxios(handlers: AxiosRequestHandler[]): Axios;
```

[mockAxios()] is a helper which simplifies mocking out of [axios] (our selected
HTTP(S) client library) in tests. To use it, you want to create `__mock__/axios.ts`,
or `__mock__/axios/index.ts` file in the root of the host codebase:
```tsx
// Basic __mock__/axios.ts example.

import { type AxiosRequestHandlerT, mockAxios } from '@dr.pogodin/react-utils/jest';

const sampleHandler: AxiosRequestHandlerT = (config) {
  if (config.url === 'https://dr.pogodin.studio/docs/react-utils') {
    return { data: 'mock reponse' };
  }
};

export default mockAxios([
  sampleHandler,
]);
```

With such setup, every call to [axios] within Jest tests will be run through
each of provided mock request handlers; if any handler returns an object,
the corresponding [axios] call will be returned to that result (with any
omitted fields of [axios] response set to reasonable default values). If all
handlers return **null** or **undefined** results, the mock will fallback
to the real network call, and it will print a warning to the console, with
details of the corresponding request and response (which will help you to
mock that out).

For further details refer to [axios] documentation for
[request config](https://www.npmjs.com/package/axios#request-config)
(the argument of mock request handlers),
and [response schema](https://www.npmjs.com/package/axios#response-schema)
(the result of mock request handlers should be a
[partial](https://www.typescriptlang.org/docs/handbook/utility-types.html#partialtype)
of it).

### mockClientSide()
```tsx
import { mockClientSide } from '@dr.pogodin/react-utils/jest';

function mockClientSide();
```
Tricks **react-utils** library into thinking the test is running within
the client-side (browser) environment.

### mockTimer()
```tsx
import { mockTimer } from '@dr.pogodin/react-utils/jest';

async function mockTimer(time: number): Promise<void>;
```
Advances mock timers, and mock date by the specified time step.

**Arguments**

- `time` - **number** - The time step [ms].

**Returns**

- **Promise** - Resolves once all async operations triggered by the mock
  time advance have completed.

### mount()
```tsx
import { mount } from '@dr.pogodin/react-utils/jest';

function mount(scene: ReactNode): MountedSceneT;
```
Mounts given `scene` into DOM and returns the root scene element (DOM node) with
**.destroy()** and **.snapshot()** methods attached. The former of these methods
unmounts the scene from DOM; and the later snapshots the scene (_i.e._ is is
an alias for `expect(scene).toMatchSnapshot()`).

**Arguments**
- `scene` - **React.ReactNode** - The scene.

**Returns**
- **HTMLElement** - The root DOM element of the mounted scene with
  **.destroy()** and **.snapshot()** methods attached.

### render()
:::danger Deprecated
Removed in **v1.34.0**. Migrate to [render()] provided by [@testing-library/react].
:::

### shallowRender()
:::danger Deprecated
Removed in **v1.34.0**. Starting with React v19 it is recommended to
[avoid shallow rendering] in tests.
:::

### shallowSnapshot()
:::danger Deprecated
Removed in **v1.34.0**. Starting with React v19 it is recommended to
[avoid shallow rendering] in tests.
:::

### simulate()
:::danger Deprecated
Removed in **v1.34.0**. Use instead [@testing-library/user-event],
or the lower-level [fireEvent] from [@testing-library/dom].
:::

### snapshot()
[snapshot()]: #snapshot
```tsx
import { snapshot } from '@dr.pogodin/react-utils/jest';

// 1st overload (synchronous):
function snapshot(element: ReactElement): RenderResult;

// 2nd overload (asynchronous):
async function snapshot(
  element: ReactElement,
  options: SnapshotOptionsT,
): Promise<RenderResult>;
```
It does the snapshot test of given React component.

<details>
<summary>Example</summary>

```tsx
/** @jest-environment jsdom */

import { snapshot } from '@dr.pogodin/react-utils/jest';

test('A snapshot test', () => {
  snapshot(<div>Example</div>);
});
```
</details>

:::caution Beware
- It relies on [render()] function provided by [@testing-library/react],
  thus it renders provided component into DOM (**appends it to the current DOM
  content**), and snapshots HTML representation of the generated DOM node.

- It requires JSDom test environment, thus a Jest test relying on this method
  should start with the magic comment `/** @jest-environment jsdom */`, or it
  should provide a virtual DOM in an alternative way.

- Prior to **v1.34.0** this method relied on now deprecated
  [react-test-renderer](https://www.npmjs.com/package/react-test-renderer),
  and thus it was snapshotting and returning a JSON representation of the rendered
  component tree, without requiring or modifying the virtual DOM.
:::

**Arguments**
- `component` &mdash; **ReactElement** &mdash; React element to snapshot.
- `options` &mdash; [SnapshotOptionsT] | **undefined** &mdash; Optional settings.
  If provided, the function will be considered asynchronous, otherwise synchronous.

**Returns**
- [RenderResult](https://testing-library.com/docs/angular-testing-library/api/#renderresult) &mdash;
  render result from [@testing-library/react].

#### SnapshotOptionsT
[SnapshotOptionsT]: #snapshotoptionst
```ts
type SnapshotOptionsT = {
  await?: Promise<void>;
};
```
Optional settings for [snapshot()] method.
- `await` &mdash; **Promise&lt;void&gt;** | **undefined** &mdash; Optional.
  If provided, [snapshot()] will await for this promise prior to capturing
  the result of component render (it is awaited after the render is triggered).

### unmockClientSide()
```tsx
import { unmockClientSide } from '@dr.pogodin/react-utils/jest';

function unmockClientSide();
```
Reverts the effect of previous [mockClientSide()] call.

[@testing-library/dom]: https://testing-library.com/docs/dom-testing-library/intro
[@testing-library/react]: https://testing-library.com/docs/react-testing-library/intro/
[@testing-library/user-event]: https://testing-library.com/docs/user-event/intro
[avoid shallow rendering]: https://react.dev/blog/2024/04/25/react-19-upgrade-guide#removed-react-test-renderer-shallow
[axios]: https://www.npmjs.com/package/axios
[E2eSsrEnv]: /docs/api/classes/E2eSsrEnv
[fireEvent]: https://testing-library.com/docs/dom-testing-library/api-events
[Jest]: https://jestjs.io
[JU]: /docs/api/utils/jest-utils
[mockClientSide()]: #mockclientside
[render()]: https://testing-library.com/docs/react-testing-library/api#render
[shallowRender()]: #shallowrender
