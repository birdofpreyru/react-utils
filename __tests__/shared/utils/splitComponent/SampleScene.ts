/**
 * @jest-environment ./src/shared/utils/jest/E2eSsrEnv.ts
 * @webpack-config-factory ./config/webpack/app-development.js
 * @webpack-config-options {
 *  "entry": "./__assets__/SampleScene",
 *  "dontUseHmr": true,
 *  "dontUseReactGlobalStateDebugging": true,
 *  "typescript": true }
 * @ssr-options { "entry": "./__assets__/SampleScene/SampleScene" }
 */

import pretty from 'pretty';
import { act } from 'react';

import { getGlobal } from 'utils/jest';

const global = getGlobal();

const outputPath = global.webpackConfig!.output!.path;

document.write(global.ssrMarkup!);

const INJ_REGEX = /<meta itemprop="drpruinj" content="[a-zA-Z0-9+/=]+">/;

let headMarkupWithoutInj: string;

it('generates expected head markup during SSR', () => {
  // TODO: Should be improved, to ensure the injection part is the same between
  // test invocations, and thus entire header can be snapshotted.
  const headMarkup = document.head.innerHTML;
  const inj = document.querySelector('meta[itemprop="drpruinj"]')!.outerHTML;
  expect(!!inj.match(INJ_REGEX)).toBe(true);
  headMarkupWithoutInj = headMarkup.replace(inj, '');
  expect(pretty(headMarkupWithoutInj)).toMatchSnapshot();
});

it('generates expected markup during SSR', () => {
  // TODO: It was done this way when document body contained server-side
  // injected data, which changed between test invocations. Now they are moved
  // to the header (see a comment above), thus this can be simplified:
  // snapshot entire body, or better ensure injected data do not change
  // between test invocations, and snapshot entire document.
  const container = document.querySelector('#react-view')!;
  expect(pretty(container.innerHTML)).toMatchSnapshot();
});

it('hydration works as expected', async () => {
  const viewMarkup = document.querySelector('#react-view')!.innerHTML;

  const fs = global.webpackOutputFs;
  const mainJs = fs.readFileSync(`${outputPath}/main.js`, 'utf8') as string;
  const aJs = fs.readFileSync(`${outputPath}/sample-component-a.js`, 'utf8') as string;
  const bJs = fs.readFileSync(`${outputPath}/sample-component-b.js`, 'utf8') as string;
  const cJs = fs.readFileSync(`${outputPath}/sample-component-c.js`, 'utf8') as string;

  /* eslint-disable @typescript-eslint/no-implied-eval, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call */
  await act(() => new Function(mainJs)());
  await act(() => new Function(aJs)());
  await act(() => new Function(bJs)());
  await act(() => new Function(cJs)());
  /* eslint-enable @typescript-eslint/no-implied-eval, @typescript-eslint/no-unsafe-return, @typescript-eslint/no-unsafe-call */

  expect(document.head.innerHTML).toBe(headMarkupWithoutInj);
  expect(document.querySelector('#react-view')!.innerHTML).toBe(viewMarkup);
});
