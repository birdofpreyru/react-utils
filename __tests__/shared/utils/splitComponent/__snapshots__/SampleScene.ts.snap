// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`generates expected head markup during SSR 1`] = `
"<title data-rh="true"></title>

<meta name="theme-color" content="#FFFFFF">

<link href="/main.css" rel="stylesheet">
<link href="/sample-component-a.css" rel="stylesheet">
<link href="/sample-component-b.css" rel="stylesheet">
<link href="/sample-component-c.css" rel="stylesheet">

<meta charset="utf-8">
<meta content="width=device-width,initial-scale=1.0" name="viewport">"
`;

exports[`generates expected markup during SSR 1`] = `
"<div class="-dr-pogodin-react-utils_____tests__-shared-utils-splitComponent-__assets__-SampleScene-style___container___oJaj7t">
  <h1>SampleCodeSplit</h1>
  <div>testValue</div><!--$-->
  <div>
    <h1>Sample Component</h1>
    <div>testValue</div>
    <h3>Prop</h3>
    <div></div>
    <h3>Children</h3>
    <div></div>
  </div><!--/$--><!--$-->
  <div>
    <h1>Sample Component</h1>
    <div>testValue</div>
    <h3>Prop</h3>
    <div>test</div>
    <h3>Children</h3>
    <div>Test</div>
  </div><!--/$--><!--$-->
  <div>
    <h1>Sample Component</h1>
    <div>testValue</div>
    <h3>Prop</h3>
    <div></div>
    <h3>Children</h3>
    <div>
      <div>a</div>
      <div>b</div>
    </div>
  </div><!--/$-->
</div>"
`;
